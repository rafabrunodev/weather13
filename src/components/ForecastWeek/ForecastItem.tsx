import moment from "moment";
import styled, { css } from "styled-components";
import { IForecast } from "../../services/weatherService";
import WeatherBlock from "../WeatherBlock";

interface IForecastItem {
  dayForecast: IForecast;
  nightForecast: IForecast;
}

interface IWeatherContainer {
  isDayTime: boolean;
}

const WeatherContainer = styled.div<IWeatherContainer>`
  min-height: 80px;
  padding: 10px 0;
  border-radius: 15px;
  display: flex;
  flex-direction: row-reverse;
  justify-content: space-between;
  align-items: center;
  color: white;

  ${({ isDayTime }) =>
    isDayTime
      ? css`
          background-color: #e89f0038;
        `
      : css`
          background-color: #041aba20;
        `}
`;

const ForecastItem: React.FC<IForecastItem> = ({
  dayForecast,
  nightForecast,
}) => {
  const isDayTime = moment().hours() < 18;

  return (
    <WeatherContainer isDayTime={isDayTime}>
      <WeatherBlock forecast={dayForecast} />
      <WeatherBlock forecast={nightForecast} />
    </WeatherContainer>
  );
};

export default ForecastItem;
