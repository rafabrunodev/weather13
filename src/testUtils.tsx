import { IForecast } from "./services/weatherService";

export const forecastListMock: IForecast[] = [
  {
    dayOfWeek: "MockDayOfWeek",
    startTime: new Date().toDateString(),
    endTime: new Date().toDateString(),
    isDaytime: true,
    temperature: 50,
    temperatureUnit: "F",
    windSpeed: "5 km",
    shortForecast: "Clear",
  },
  {
    dayOfWeek: "MockDayOfWeekNight",
    startTime: new Date().toDateString(),
    endTime: new Date().toDateString(),
    isDaytime: false,
    temperature: 10,
    temperatureUnit: "F",
    windSpeed: "2 km",
    shortForecast: "With Fog",
  },
];

export const forecastMockAPIResponse = {
  properties: {
    periods: [
      {
        number: 1,
        name: "This Afternoon",
        startTime: "2023-02-07T12:00:00-06:00",
        endTime: "2023-02-07T18:00:00-06:00",
        isDaytime: true,
        temperature: 48,
        temperatureUnit: "F",
        temperatureTrend: null,
        windSpeed: "5 to 10 mph",
        windDirection: "N",
        icon: "https://api.weather.gov/icons/land/day/skc?size=medium",
        shortForecast: "Sunny",
        detailedForecast: "Sunny, with a high near 48. North wind 5 to 10 mph.",
      },
      {
        number: 2,
        name: "Tonight",
        startTime: "2023-02-07T18:00:00-06:00",
        endTime: "2023-02-08T06:00:00-06:00",
        isDaytime: false,
        temperature: 26,
        temperatureUnit: "F",
        temperatureTrend: null,
        windSpeed: "5 mph",
        windDirection: "SW",
        icon: "https://api.weather.gov/icons/land/night/few?size=medium",
        shortForecast: "Mostly Clear",
        detailedForecast:
          "Mostly clear, with a low around 26. Southwest wind around 5 mph.",
      },
    ],
  },
};

export const forecastMockGrid = {
  gridId: "MOCK_ID",
  gridX: "MOCK_X",
  gridY: "MOCK_Y",
};
