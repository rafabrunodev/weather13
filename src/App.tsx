import { useState, createContext, useEffect } from "react";
import styled from "styled-components";
import ForecastCover from "./components/ForecastCover";
import Header from "./components/Header";
import ForecastWeek from "./components/ForecastWeek";

import { getForecastList, IForecast } from "./services/weatherService";
import { device } from "./utils";

export const ForecastContext = createContext<IForecast[]>([]);

const AppContainer = styled.div`
  text-align: center;
  margin: 0 auto;
  width: 100%;
`;

const AppContent = styled.div`
  display: flex;
  flex-direction: column;

  @media ${device.laptop} {
    flex-direction: row;
    justify-content: center;
    align-items: flex-start;
  }
`;

function App() {
  const [forecast, setForecast] = useState<IForecast[]>([]);
  const [loading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    const getApi = async () => {
      const forecastList = await getForecastList();

      if (forecastList) {
        setLoading(true);
        setForecast(forecastList);
      }
    };

    getApi();
  }, []);

  return (
    <AppContainer>
      <Header />
      <ForecastContext.Provider value={forecast}>
        {loading ? (
          <AppContent>
            <ForecastCover />
            <ForecastWeek />
          </AppContent>
        ) : (
          <h3>Loading...</h3>
        )}
      </ForecastContext.Provider>
    </AppContainer>
  );
}

export default App;
