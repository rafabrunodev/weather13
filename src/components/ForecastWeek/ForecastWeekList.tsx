import styled from "styled-components";
import { IForecast } from "../../services/weatherService";
import { device } from "../../utils";
import ForecastItem from "./ForecastItem";

interface IForecastWeekList {
  forecastByDay: IForecast[][];
}

const Ul = styled.ul`
  list-style: none;
  width: 100%;
  margin: 0;
  padding: 0;

  @media ${device.laptop} { 
    width: 100%;
    padding: 0 40px 30px 40px;
  }
`;

const Li = styled.li`
  margin: 10px;
`;

const DayOfWeek = styled.h4`
  text-align: left;

  small {
    font-weight: 400;
    font-size: 12px;
  }
`;

const ForecastWeekList: React.FC<IForecastWeekList> = ({ forecastByDay }) => {
  return (
    <Ul>
      {forecastByDay.map((forecast, index) => (
        <Li key={index}>
          <DayOfWeek>{index === 0 ? "Today" : forecast[0].dayOfWeek}</DayOfWeek>
          <ForecastItem dayForecast={forecast[0]} nightForecast={forecast[1]} />
        </Li>
      ))}
    </Ul>
  );
};

export default ForecastWeekList;
