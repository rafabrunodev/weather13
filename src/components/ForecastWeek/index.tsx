import  ForecastWeek from './ForecastWeek'

export default ForecastWeek;