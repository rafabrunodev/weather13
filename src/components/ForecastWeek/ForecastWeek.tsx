import { useContext } from "react";
import { ForecastContext } from "../../App";
import { IForecast } from "../../services/weatherService";
import ForecastWeekList from "./ForecastWeekList";

const getForecastByDay = (forecastList: IForecast[]) => {
  const chunkSize = 2;
  const myArray = [];

  for (let i = 0; i < forecastList.length; i += chunkSize) {
    myArray.push(forecastList.slice(i, i + chunkSize));
  }

  return myArray;
};

const ForecastWeek: React.FC = () => {
  const forecastList: IForecast[] = useContext(ForecastContext);
  const forecastByDay = getForecastByDay(forecastList);
  return <ForecastWeekList forecastByDay={forecastByDay} />;
};

export default ForecastWeek;
