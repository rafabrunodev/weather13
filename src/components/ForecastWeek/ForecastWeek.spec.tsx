import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import ForecastWeek from "./ForecastWeek";
import { forecastListMock } from "../../testUtils";
import { ForecastContext } from "../../App";

vi.mock("moment", async () => {
  const actual: any = await vi.importActual("moment");
  return {
    ...actual,
    hours: () => 21
  }
})

describe("ForecastWeek.tsx", () => {
  it("should match Snapshot", () => {
    const { container } = render(
      <ForecastContext.Provider value={forecastListMock}>
        <ForecastWeek />
      </ForecastContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
