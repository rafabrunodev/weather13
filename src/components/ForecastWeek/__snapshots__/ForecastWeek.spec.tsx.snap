// Vitest Snapshot v1

exports[`ForecastWeek.tsx > should match Snapshot 1`] = `
<div>
  <ul
    class="sc-ksBlkl jqMxVU"
  >
    <li
      class="sc-hBxehG rfweg"
    >
      <h4
        class="sc-fnGiBr eJWxAO"
      >
        Today
      </h4>
      <div
        class="sc-bqWxrE cCwAsz"
      >
        <div
          class="sc-csuSiG cCWVKd"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="40"
            role="dayIcon"
            size="40"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 30 30"
            width="40"
          >
            <path
              d="M4.37,14.62c0-0.24,0.08-0.45,0.25-0.62c0.17-0.16,0.38-0.24,0.6-0.24h2.04c0.23,0,0.42,0.08,0.58,0.25c0.15,0.17,0.23,0.37,0.23,0.61S8,15.06,7.85,15.23c-0.15,0.17-0.35,0.25-0.58,0.25H5.23c-0.23,0-0.43-0.08-0.6-0.25C4.46,15.06,4.37,14.86,4.37,14.62zM7.23,21.55c0-0.23,0.08-0.43,0.23-0.61l1.47-1.43c0.15-0.16,0.35-0.23,0.59-0.23c0.24,0,0.44,0.08,0.6,0.23s0.24,0.34,0.24,0.57c0,0.24-0.08,0.46-0.24,0.64L8.7,22.14c-0.41,0.32-0.82,0.32-1.23,0C7.31,21.98,7.23,21.78,7.23,21.55zM7.23,7.71c0-0.23,0.08-0.43,0.23-0.61C7.66,6.93,7.87,6.85,8.1,6.85c0.22,0,0.42,0.08,0.59,0.24l1.43,1.47c0.16,0.15,0.24,0.35,0.24,0.59c0,0.24-0.08,0.44-0.24,0.6s-0.36,0.24-0.6,0.24c-0.24,0-0.44-0.08-0.59-0.24L7.47,8.32C7.31,8.16,7.23,7.95,7.23,7.71zM9.78,14.62c0-0.93,0.23-1.8,0.7-2.6s1.1-1.44,1.91-1.91s1.67-0.7,2.6-0.7c0.7,0,1.37,0.14,2.02,0.42c0.64,0.28,1.2,0.65,1.66,1.12c0.47,0.47,0.84,1.02,1.11,1.66c0.27,0.64,0.41,1.32,0.41,2.02c0,0.94-0.23,1.81-0.7,2.61c-0.47,0.8-1.1,1.43-1.9,1.9c-0.8,0.47-1.67,0.7-2.61,0.7s-1.81-0.23-2.61-0.7c-0.8-0.47-1.43-1.1-1.9-1.9C10.02,16.43,9.78,15.56,9.78,14.62zM11.48,14.62c0,0.98,0.34,1.81,1.03,2.5c0.68,0.69,1.51,1.04,2.49,1.04s1.81-0.35,2.5-1.04s1.04-1.52,1.04-2.5c0-0.96-0.35-1.78-1.04-2.47c-0.69-0.68-1.52-1.02-2.5-1.02c-0.97,0-1.8,0.34-2.48,1.02C11.82,12.84,11.48,13.66,11.48,14.62zM14.14,22.4c0-0.24,0.08-0.44,0.25-0.6s0.37-0.24,0.6-0.24c0.24,0,0.45,0.08,0.61,0.24s0.24,0.36,0.24,0.6v1.99c0,0.24-0.08,0.45-0.25,0.62c-0.17,0.17-0.37,0.25-0.6,0.25s-0.44-0.08-0.6-0.25c-0.17-0.17-0.25-0.38-0.25-0.62V22.4zM14.14,6.9V4.86c0-0.23,0.08-0.43,0.25-0.6C14.56,4.09,14.76,4,15,4s0.43,0.08,0.6,0.25c0.17,0.17,0.25,0.37,0.25,0.6V6.9c0,0.23-0.08,0.42-0.25,0.58S15.23,7.71,15,7.71s-0.44-0.08-0.6-0.23S14.14,7.13,14.14,6.9zM19.66,20.08c0-0.23,0.08-0.42,0.23-0.56c0.15-0.16,0.34-0.23,0.56-0.23c0.24,0,0.44,0.08,0.6,0.23l1.46,1.43c0.16,0.17,0.24,0.38,0.24,0.61c0,0.23-0.08,0.43-0.24,0.59c-0.4,0.31-0.8,0.31-1.2,0l-1.42-1.42C19.74,20.55,19.66,20.34,19.66,20.08zM19.66,9.16c0-0.25,0.08-0.45,0.23-0.59l1.42-1.47c0.17-0.16,0.37-0.24,0.59-0.24c0.24,0,0.44,0.08,0.6,0.25c0.17,0.17,0.25,0.37,0.25,0.6c0,0.25-0.08,0.46-0.24,0.62l-1.46,1.43c-0.18,0.16-0.38,0.24-0.6,0.24c-0.23,0-0.41-0.08-0.56-0.24S19.66,9.4,19.66,9.16zM21.92,14.62c0-0.24,0.08-0.44,0.24-0.62c0.16-0.16,0.35-0.24,0.57-0.24h2.02c0.23,0,0.43,0.09,0.6,0.26c0.17,0.17,0.26,0.37,0.26,0.6s-0.09,0.43-0.26,0.6c-0.17,0.17-0.37,0.25-0.6,0.25h-2.02c-0.23,0-0.43-0.08-0.58-0.25S21.92,14.86,21.92,14.62z"
            />
          </svg>
          <div
            class="sc-eDWCr jfUuUf"
          >
            <strong>
              50
              °
              <small>
                F
              </small>
            </strong>
            <span>
              Clear
            </span>
          </div>
        </div>
        <div
          class="sc-csuSiG cCWVKd"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="40"
            role="nightIcon"
            size="40"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 30 30"
            width="40"
          >
            <path
              d="M4.14,16.9c0-1.16,0.35-2.18,1.06-3.08s1.62-1.47,2.74-1.72c0.23-1.03,0.7-1.93,1.4-2.7c0.7-0.77,1.55-1.32,2.53-1.65c0.62-0.21,1.26-0.32,1.93-0.32c0.81,0,1.6,0.16,2.35,0.48c0.28-0.47,0.61-0.88,0.99-1.22c0.38-0.34,0.77-0.61,1.17-0.79c0.4-0.18,0.8-0.32,1.18-0.41s0.76-0.13,1.12-0.13c0.38,0,0.79,0.05,1.23,0.16l0.82,0.25c0.14,0.06,0.18,0.13,0.14,0.22l-0.14,0.6c-0.07,0.31-0.1,0.6-0.1,0.86c0,0.31,0.05,0.63,0.15,0.95c0.1,0.32,0.24,0.63,0.44,0.94c0.19,0.31,0.46,0.58,0.8,0.83c0.34,0.25,0.72,0.44,1.15,0.57l0.62,0.22c0.1,0.03,0.15,0.08,0.15,0.16c0,0.02-0.01,0.04-0.02,0.07l-0.18,0.67c-0.27,1.08-0.78,1.93-1.5,2.57c0.4,0.7,0.62,1.45,0.65,2.24c0.01,0.05,0.01,0.12,0.01,0.23c0,0.89-0.22,1.72-0.67,2.48c-0.44,0.76-1.05,1.36-1.8,1.8c-0.76,0.44-1.59,0.67-2.48,0.67H9.07c-0.89,0-1.72-0.22-2.48-0.67s-1.35-1.05-1.79-1.8S4.14,17.8,4.14,16.9zM5.85,16.9c0,0.89,0.32,1.66,0.96,2.31c0.64,0.65,1.39,0.98,2.26,0.98h10.81c0.89,0,1.65-0.32,2.28-0.97s0.95-1.42,0.95-2.32c0-0.88-0.32-1.63-0.96-2.26c-0.64-0.63-1.4-0.95-2.28-0.95h-1.78l-0.1-0.75c-0.1-1.01-0.52-1.88-1.26-2.59s-1.62-1.11-2.63-1.2c-0.03,0-0.08,0-0.15-0.01c-0.07-0.01-0.11-0.01-0.15-0.01c-0.51,0-1.02,0.1-1.54,0.29V9.4c-0.73,0.28-1.35,0.74-1.84,1.37c-0.5,0.63-0.8,1.35-0.9,2.17l-0.07,0.72l-0.68,0.03c-0.84,0.1-1.54,0.45-2.1,1.06S5.85,16.07,5.85,16.9zM17.6,8.79c1.06,0.91,1.72,1.97,1.97,3.18h0.32c1.24,0,2.3,0.39,3.17,1.18c0.33-0.31,0.58-0.67,0.76-1.07c-0.91-0.43-1.63-1.09-2.16-1.97c-0.52-0.88-0.79-1.81-0.79-2.78V7.09c-0.05-0.01-0.13-0.01-0.24-0.01c-0.58-0.01-1.15,0.13-1.7,0.44C18.38,7.82,17.93,8.24,17.6,8.79z"
            />
          </svg>
          <div
            class="sc-eDWCr jfUuUf"
          >
            <strong>
              10
              °
              <small>
                F
              </small>
            </strong>
            <span>
              With Fog
            </span>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
